crates/but-worktrees/src/integrate.rs

```diff
@@ -0,0 +1,278 @@
+use gitbutler_oxidize::{GixRepositoryExt as _, ObjectIdExt};
+use gitbutler_project::access::{WorktreeReadPermission, WorktreeWritePermission};
+use gitbutler_stack::{Stack, VirtualBranchesHandle};
+use gitbutler_workspace::branch_trees::{
+    WorkspaceState, merge_workspace, move_tree, update_uncommited_changes,
+};
+use gix::prelude::ObjectIdExt as _;
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    db::{delete_worktree_meta, get_worktree_meta},
+    git::git_worktree_remove,
+};
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+#[serde(tag = "type", content = "data", rename_all = "camelCase")]
+/// This gets used as a public API in the CLI so be careful when modifying.
+pub enum WorktreeIntegrationStatus {
+    NoMergeBaseFound,
+    WorktreeIsBare,
+    /// If we were to integrate this worktree back into the project, it would
+    /// cause the workspace to conflict.
+    ///
+    /// If this is true, the worktree can't be integrated.
+    CausesWorkspaceConflicts,
+    Integratable {
+        /// The cherry pick produced when integrating will be conflicted.
+        cherry_pick_conflicts: bool,
+        /// Commits above where this worktree will be cherry-picked are going to
+        /// end up conflicted.
+        commits_above_conflict: bool,
+        /// Whether the uncommitted changes in the main checkout will end up
+        /// conflicted
+        working_dir_conflicts: bool,
+    },
+}
+
+/// Determines whether a worktree is integrated
+///
+/// This function makes use of older APIs because there is not yet an
+/// alternative to the rebase engine.
+pub fn worktree_integration_status(
+    ctx: &mut CommandContext,
+    perm: &WorktreeReadPermission,
+    path: &Path,
+    target: &gix::refs::FullNameRef,
+) -> Result<WorktreeIntegrationStatus> {
+    Ok(worktree_integration_inner(ctx, perm, path, target)?.0)
+}
+
+/// Integrates a worktree if it's integratable
+///
+/// This function makes use of older APIs because there is not yet an
+/// alternative to the rebase engine.
+pub fn worktree_integrate(
+    ctx: &mut CommandContext,
+    perm: &mut WorktreeWritePermission,
+    path: &Path,
+    target: &gix::refs::FullNameRef,
+) -> Result<()> {
+    let before = WorkspaceState::create(ctx, perm.read_permission())?;
+
+    let result = worktree_integration_inner(ctx, perm.read_permission(), path, target)?;
+    let (WorktreeIntegrationStatus::Integratable { .. }, Some(mut status)) = result else {
+        bail!("Worktree failed integration checks");
+    };
+
+    status
+        .stack
+        .set_heads_from_rebase_output(ctx, status.rebase_output.references)?;
+    let after = WorkspaceState::create(ctx, perm.read_permission())?;
+    update_uncommited_changes(ctx, before, after, perm)?;
+    let vb_state = VirtualBranchesHandle::new(ctx.project().gb_dir());
+    update_workspace_commit(&vb_state, ctx)?;
+
+    git_worktree_remove(&ctx.project().path, path, true)?;
```

Comment:
Worktrees *really* must not be identified by `path`, but by their [id](https://docs.rs/gix/latest/gix/struct.Worktree.html#method.id). This also exists on the worktree proxies that are iterated: [`id`](https://docs.rs/gix/latest/gix/worktree/struct.Proxy.html#method.id).